guacsec/guac
# Create a GraphQL interface for backend

[See design doc](https://docs.google.com/document/d/1yZ3-ZcfnRDWgw9uZlPuLmIHS9pNMr3DO_AEbHsDXmN8/edit?usp=sharing)

I separated the PR into smaller commits so that each can be reviewed individually since this is quite large due to autogenerated code.

This is going to be draft, will add more work on PRs on top of it until the interface stabilizes.

To test:
- clone this branch
- start the neo4j database, ingest data as in `SETUP.md`
- run the demo server: `go run pkg/assembler/graphql/server.go`
- on the GraphQL playground run this query (the only one supported right now)
```
query artifacts {
  artifacts {
    name
    digest
  }
}
```
- in `pkg/assembler/graphql/server.go` can change backend (`s/true/false` and the other way round) and test again with the other one.

Left to do:
- a large number of TODOs in the schema and code
- also add a client side interface, try to replicate the certifier usecase
- add more query types
- redefine nodes and artifacts as needed
- migrate existing code to use the graphql interface
- fix all checks in CI. Left for later to focus first on the interface

*Note*: to make sure graphql and existing assembler are kept in sync (and don't get updates which will be impossible to handle on the other side), I added a `.CODEOWNERS` to automatically request review from me whenever stuff under `/pkg/assembler` changes
